kube-runtime/src/coordination.rs

```diff
@@ -0,0 +1,603 @@
+            ));
+        }
+        if self.retry_period.as_secs() < 1 {
+            return Err(Error::ConfigError(
+                "retry_period must be at least 1 second".into(),
+            ));
+        }
+        if self.api_timeout.as_secs() < 1 {
+            return Err(Error::ConfigError("api_timeout must be at least 1 second".into()));
+        }
+        Ok(Config(self))
+    }
+}
+
+/// A task which is responsible for acquiring and maintaining a `coordination.k8s.io/v1` `Lease`
+/// to establish leadership.
+pub struct LeaderElector {
+    /// A K8s API wrapper around the client.
+    api: Api<Lease>,
+    /// Leader election config.
+    config: ConfigBuilder,
+    /// The internal state of this task.
+    state: State,
+    /// The state signal, which always reflects the current internal state of this task.
+    state_tx: watch::Sender<LeaderState>,
+    /// Shutdown channel.
+    shutdown: oneshot::Receiver<()>,
+    /// A bool indicating that there was an error encountered on the last attempt to acquire the lease.
+    ///
+    /// This is used as a simple retry / backoff indicator.
+    had_error_on_last_try: bool,
+}
+
+impl LeaderElector {
+    /// Create a new `LeaderElector` instance & spawn it onto the runtime for execution.
+    #[must_use = "handle must be used for observing state changes and graceful shutdown"]
+    pub fn spawn(config: Config, client: Client) -> LeaderElectorHandle {
+        let (state_tx, state_rx) = watch::channel(LeaderState::Standby);
+        let (shutdown_tx, shutdown_rx) = oneshot::channel();
+        let this = LeaderElector {
+            api: Api::namespaced(client, &config.0.namespace),
+            config: config.0,
+            state_tx,
+            state: State::Standby,
+            shutdown: shutdown_rx,
+            had_error_on_last_try: false,
+        };
+        let handle = tokio::spawn(this.run());
+        LeaderElectorHandle {
+            shutdown: shutdown_tx,
+            state: state_rx,
+            handle,
+        }
+    }
+
+    async fn run(mut self) {
+        tracing::info!("leader elector task started");
+
+        // Perform an initail pass at acquiring / renewing the lease.
+        if let Err(err) = self.try_acquire_or_renew().await {
+            tracing::error!(error = ?err, "error attempting to acquire/renew lease");
+        }
+        tracing::info!("finished initial call to try_acquire_or_renew");
+
+        let lease_watcher = watcher(
+            self.api.clone(),
+            ListParams {
+                field_selector: Some(format!("metadata.name={}", self.config.name)),
+                ..Default::default()
+            },
+        );
+        tokio::pin!(lease_watcher);
+
+        loop {
+            let delay_duration = self.get_next_acquire_renew_time();
+            tracing::debug!("delaying for {}s", delay_duration.as_secs());
+            let delay = tokio::time::sleep(delay_duration);
```

Comment:
> If the object we are watching updates more frequently than the delay_duration here, I think we risk ending up always processing outdated lease events in the select.
> [...]
> I still think we can end up starving the stream here though and get into trouble based on how we just use the event straight inside handle_lease_event though.

Though I believe this is clear, I will clarify that the `delay_duration` has no bearing over the event stream. As lease events come in, they are immediately processed, and `delay_duration` does not hinder that at all.

You can see in the code that the `select!` will be triggered on whichever is first, either a lease event, or if no lease events come in then the delay branch will be triggered once its timer has elapsed. The delay timer does not block the event stream.

You can see that in either case, as soon as processing of the branch finishes, control will loop and immediately begin awaiting the next lease event or the next timeout.

### Starvation
There are actually no ways for the event stream to be starved here. Starvation in this context means that the event stream is never processed as other work is being prioritized.

Well, in this case, there is only 1 other thing which could be worked on: the delay timer. How is this timer constructed? Via `get_next_acquire_renew_time()`, and the following is a high-level description of the algorithm:
- When leader, we use the `renew_deadline` (recommended 10s) to establish the delay. The delay will ONLY be 0s if we need to renew now, else it will be the delta until we need to delay.
  - I will point out here that in a properly running system, no other controllers will be updating the lease, and the leader will be periodically updating the lease (heartbeat) according to the renew deadline. This yields a system which is really just a timer, and a single update on the renew interval.
- When follower, the algorithm is a bit more complex, as we compensate for potential clock drift by only using our own process clock, and we reckon updates only based on when we observe them, not according to their recorded time on the lease (this is what the upstream Go / core controllers do).
  - Additionally, we include a jitter factor, and followers will only contend for the lease when `lease_duration` has elapsed, which is validated to be longer than the leader's `renew_deadline`.
  - So, as long as a leader is indeed heartbeating the lease, even in the face of arbitrary clock drift, the system will remain robust with a stable leader.
- When the lease is pristine (neither leader nor follower), then 0, as we will contend to create & acquire the lease.

### Conclusion
As can be seen above and in the code, the algorithm's timers will not starve the event stream. The event stream itself is rather minimal and robust, and even if a large number of events were observed back to back, the system will be able to stay up-to-date.

### Concession
If the above is still not satisfactory, then there is a very simple solution: any time we process an event, first attempt to drain the entire stream via try_poll, and only process the most recent event.

Reply:
@clux upon re-reading your comment here, I believe that this is a simple misunderstanding. The `delay_duration` does not block processing. It is a fallback ONLY for cases where no lease updates have been observed for a period longer than the delay.

One way of thinking about this is as follows: the current lease holder has died (a different pod), and thus no updates are taking place on the lease. Once `delay_duration` is elapsed, that is our cue to go and attempt to acquire leadership.